scripts/_functions_sass.R

```diff
@@ -0,0 +1,451 @@
+  path_out <- ""
+
+  cli::cli_progress_step(
+    "{verb}ing {.strong {preset}} theme CSS...",
+    "{verb}ed {.strong {preset}} theme CSS {.path {path_out}}",
+    "Failed to precompile {.strong {preset}} theme CSS."
+  )
+
+  # Pre-render the Sass files into compiled CSS
+  tryCatch(
+    {
+      sass(
+        sass_file(path_preset_scss),
+        output = path_preset_compiled,
+        options = sass_options(
+          output_style = "compressed",
+          source_comments = FALSE,
+          source_map_embed = FALSE
+        )
+      )
+      if (preset %in% presets_precompiled) {
+        path_out <- path_rel(path_preset_compiled)
+      }
+      cli::cli_progress_done()
+    },
+    error = function(cnd) {
+      cli::cli_progress_done(result = "error")
+      cli::cli_inform("Failed to compile Sass file {.path {path_preset_scss}}.", parent = cnd)
+    }
+  )
+
+  # Remove intermediate entry-point Sass file
+  file_delete(path_preset_scss)
+
+  # Don't bundle precompiled Bootswatch files
+  if (!preset %in% presets_precompiled) {
+    file_delete(path_preset_compiled)
+    return(invisible())
+  }
+
+  invisible(path_preset_compiled)
+}
+
+#' Write the Python preset choices to `_theme_presets.py`.
+#' @param presets A character vector of preset names.
+#' @param presets_precompiled A character vector of preset names that were precompiled.
+write_python_preset_choices <- function(presets, presets_precompiled) {
+  path_presets_py <- path_root("shiny", "ui", "_theme_presets.py")
+  cli::cli_progress_step("Generate {.path {path_rel(path_presets_py)}}")
+
+  template <- r"(# Generated by scripts/htmlDependencies.R: do not edit by hand
+
+from __future__ import annotations
+
+from typing import Literal
+
+ShinyThemePreset = Literal[
+%s
+]
+
+shiny_theme_presets: tuple[ShinyThemePreset, ...] = (
+%s
+)
+
+shiny_theme_presets_bundled: tuple[ShinyThemePreset, ...] = (
+%s
+))"
+
+  py_lines <- function(x) {
+    x <- paste(sprintf('"%s"', x), collapse = ",\n    ")
+    paste0("    ", x, ",")
+  }
+
+  writeLines(
+    sprintf(template, py_lines(presets), py_lines(presets), py_lines(presets_precompiled)),
+    path_presets_py
+  )
+
+  invisible(path_presets_py)
+}
+
+#' Copy the compiled CSS from the Shiny preset to the base Bootstrap CSS file. It also
+#' replaces the `@import` statement in the Bootstrap CSS file with the correct path to
+#' the font CSS file to re-enable pre-downloaded fonts.
+copy_shiny_preset_to_base_bootstrap <- function() {
+  path_preset_shiny <- path_root("shiny", "www", "shared", "sass", "preset", "shiny", "preset.min.css")
+  path_bootstrap <- path_root("shiny", "www", "shared", "bootstrap", "bootstrap.min.css")
+  cli::cli_progress_step("Copy shiny preset to {.path {path_rel(path_bootstrap)}}")
+
+  preset_shiny_lines <- readLines(path_preset_shiny)
+  preset_shiny_lines[1] <- sub(
+    "^@import url\\([^)]+\\)",
+    "@import url(\"font.css\")",
+    preset_shiny_lines[1]
+  )
```

Comment:
I'm having a hard time following why we need to remove `'font.css'` then add it back in. Also, this `sub()` could break if the `@import` location changes? Would it be better to only alter `$web-font-path` in the cases where we need to (instead of adding this back in after compiling)?

Reply:
To give some context: `shiny/www/shared/bootstrap/` is copied directly from bslib, where `bootstrap.min.css` is pre-compiled with the downloaded fonts but only contains Bootstrap styles. That `bootstrap.min.css` doesn't contain all of the styles we need since this PR adds the other built-in components to `bootstrap.min.css`.

It seemed easier to copy the already compiled shiny preset `bootstrap.min.css` and to replace the font import with the `font.css` import. But I agree it's potentially brittle. It's also not quite as simple as changing `$web-font-path` in one place, we'd instead need to start from `bs_full_theme()` and then recompile without going through the `bs_theme` --> sass files --> compiled css steps that the precompiled presets use.

The tradeoff I see here is to have slightly more brittle code but with two very close branches (the precompiled default theme is certainly identical to the preset theme, except for the `font.css` line), or to have less brittle code but distant branches (the precompiled default theme is not compiled from the Sass used by `ui.Theme("shiny")`).

Reply:
I added a simple check to throw an error if our assumption about that first line changes. 21fa5e9c0933b59a0bac53670842733a0d4b7535